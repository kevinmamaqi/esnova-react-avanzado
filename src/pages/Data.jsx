import { useQuery } from "@tanstack/react-query"
import React, { useMemo, useState } from "react"
import { Body, SubHeaderStyled } from "../components/layout"
import EsnovaTable from "../components/organisms/EsnovaTable"
import { apiUrls } from "../constants"
import { groupByDistricts } from "../helpers/groupByDistricts"
import { Container, FlexBox } from "../styles"

const setPriceColor = (price) => {
  if (price < 250000) return "green"
  if (price > 600000) return "red"
  return "black"
}

function Precio({ price }) {
  return <span style={{ color: setPriceColor(price) }}>{price} €</span>
}

function Data() {
  const [view, setView] = useState("viviendas")
  const { data } = useQuery({
    queryKey: ["homes"],
    queryFn: () => fetch(apiUrls.pisos).then((res) => res.json()),
  })

  const columns = useMemo(
    () => [
      { id: "title", label: "Nombre", showSortBy: true },
      {
        id: "price",
        label: "Precio",
        Cell: Precio,
      },
      {
        id: "city",
        label: "Ciudad",
        isHidden: true,
      },
      {
        id: "type",
        label: "Tipo",
        showSortBy: true,
      },
    ],
    [],
  )

  const columnsDistrict = useMemo(
    () => [
      { id: "district", label: "Barrio", showSortBy: true },
      {
        id: "numHomes",
        label: "Número de viviendas",
      },
    ],
    [],
  )

  const dataDistrict = useMemo(() => {
    if (!data) return []
    return groupByDistricts(data)
  }, [data])

  return (
    <Body>
      <SubHeaderStyled>
        <h1>Data</h1>
      </SubHeaderStyled>
      <Container>
        <FlexBox style={{ marginTop: "1rem", marginBottom: "1rem" }}>
          <button
            type="button"
            onClick={() =>
              view === "barrios" ? setView("viviendas") : setView("barrios")
            }
          >
            {view}
          </button>
        </FlexBox>
        {view === "viviendas" ? (
          <EsnovaTable columns={columns} data={data} />
        ) : (
          <EsnovaTable columns={columnsDistrict} data={dataDistrict} />
        )}
      </Container>
    </Body>
  )
}

export default Data
