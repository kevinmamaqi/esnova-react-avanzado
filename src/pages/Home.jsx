import React from "react"
import { ErrorBoundary } from "../components/atoms/ErrorBoundary"
import { Body } from "../components/layout"
import SubHeader from "../components/layout/SubHeader"
import { Homes } from "../components/organisms"
import { Container } from "../styles"

const ErrorComponent = () => {
  throw new Error("Error")
  return "hay un error"
}

const WithEBoundary = () => (
  <ErrorBoundary>
    <ErrorComponent />
  </ErrorBoundary>
)

function Home() {
  return (
    <Body>
      <WithEBoundary />
      <SubHeader />
      <Container>
        <Homes />
      </Container>
    </Body>
  )
}

export default Home
