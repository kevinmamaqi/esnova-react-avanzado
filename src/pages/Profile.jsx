import React, { useState, useMemo, useCallback, useEffect } from "react"
import styled from "styled-components"

const expensiveFunction = (color) => {
  let i = 0
  while (i < 2000000000) {
    i += 1
  }
  return color
}

const Circle = styled.div`
  width: 100px;
  height: 100px;
  border-radius: 50%;
  background-color: ${(props) => props.color};
`

function CircleComponent({ color }) {
  const colorMemo = useMemo(() => expensiveFunction(color), [color])
  return <Circle color={colorMemo} />
}

function Profile() {
  const [color, setColor] = useState("red")
  // counter
  const [counter, setCounter] = useState(0)
  const changeColor = () => {
    setColor("blue")
  }

  const diHola = useCallback(() => {
    console.log("hola")
  }, [color])

  useEffect(() => {
    window.addEventListener("click", diHola)
  }, [counter])

  return (
    <div>
      <CircleComponent color={color} />
      <button type="button" onClick={changeColor}>
        Change color
      </button>
      <br />
      {counter}
      <button type="button" onClick={() => setCounter(counter + 1)}>
        Increment
      </button>
    </div>
  )
}

export default Profile
