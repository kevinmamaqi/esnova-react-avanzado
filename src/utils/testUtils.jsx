/* eslint-disable import/no-extraneous-dependencies */
import { QueryClient, QueryClientProvider } from "@tanstack/react-query"
import { render } from "@testing-library/react"
import { BrowserRouter } from "react-router-dom"

const queryClient = new QueryClient({
  defaultOptions: {
    queries: {
      // âœ… turns retries off
      retry: false,
    },
  },
})

function AllTheProviders({ children }) {
  return (
    <QueryClientProvider client={queryClient}>
      <BrowserRouter>{children}</BrowserRouter>
    </QueryClientProvider>
  )
}

const customRender = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

export * from "@testing-library/react"
export { customRender as render }
