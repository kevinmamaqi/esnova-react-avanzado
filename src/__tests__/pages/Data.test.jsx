import { render, screen } from "../../utils/testUtils"
import { Data } from "../../pages"

describe("Data", () => {
  it("should render properly", () => {
    render(<Data />)
    expect(screen.getAllByText(/Data/)).toHaveLength(2)
  })

  it("should render requested data", async () => {
    render(<Data />)
    expect(await screen.findByText(/Piso 1/)).toBeInTheDocument()
  })
})
