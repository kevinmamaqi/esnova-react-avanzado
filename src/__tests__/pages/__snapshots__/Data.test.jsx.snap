// Vitest Snapshot v1

exports[`Data > should match snapshot 1`] = `
<div>
  <div
    class="sc-bcXHqe sc-gswNZR sc-ipEyDJ lakRFM dcrRKX jSbNFb"
    direction="row"
  >
    <h1
      class="sc-eDvSVe jZisDL"
    >
      MIPISO.com
    </h1>
    <ul
      class="sc-jrcTuL fXPeUQ sc-kDvujY"
    >
      <li
        class="sc-eDvSVe jZisDL"
      >
        <a
          href="/"
        >
          Home
        </a>
      </li>
      <li
        class="sc-eDvSVe jZisDL"
      >
        <a
          href="/data"
        >
          Data
        </a>
      </li>
      <li
        class="sc-eDvSVe jZisDL"
      >
        <a
          href="/profile"
        >
          Profile
        </a>
      </li>
    </ul>
  </div>
  <div
    class="sc-bcXHqe sc-gswNZR sc-bqWxrE leBCrZ dcrRKX kdzNdp"
  >
    <h1>
      Data
    </h1>
  </div>
  <div
    class="sc-bcXHqe sc-gswNZR leBCrZ dcrRKX"
  >
    <div
      class="sc-bcXHqe leBCrZ"
      style="margin-top: 1rem; margin-bottom: 1rem;"
    >
      <button
        type="button"
      >
        viviendas
      </button>
    </div>
    <table
      class="sc-fEXmlR jXsuTM"
    >
      <thead>
        <tr>
          <th
            class="sc-bjfHbI hnoYfZ"
          >
            <div
              class="sc-bcXHqe lakRFM"
              direction="row"
            >
              Nombre
              <div
                class="sc-hHTYSt fzXdxU"
              >
                <svg
                  class="sc-pyfCe"
                  height="6"
                  width="12"
                >
                  <polygon
                    fill="#333"
                    points="6,0 0,6 12,6"
                  />
                  Sorry, your browser does not support inline SVG.
                </svg>
                <svg
                  class="sc-pyfCe"
                  height="6"
                  width="12"
                >
                  <polygon
                    fill="#333"
                    points="0,0 12,0 6,6"
                  />
                  Sorry, your browser does not support inline SVG.
                </svg>
              </div>
            </div>
          </th>
          <th
            class="sc-bjfHbI hnoYfZ"
          >
            Precio
          </th>
          <th
            class="sc-bjfHbI hnoYfZ"
          >
            <div
              class="sc-bcXHqe lakRFM"
              direction="row"
            >
              Tipo
              <div
                class="sc-hHTYSt fzXdxU"
              >
                <svg
                  class="sc-pyfCe"
                  height="6"
                  width="12"
                >
                  <polygon
                    fill="#333"
                    points="6,0 0,6 12,6"
                  />
                  Sorry, your browser does not support inline SVG.
                </svg>
                <svg
                  class="sc-pyfCe"
                  height="6"
                  width="12"
                >
                  <polygon
                    fill="#333"
                    points="0,0 12,0 6,6"
                  />
                  Sorry, your browser does not support inline SVG.
                </svg>
              </div>
            </div>
          </th>
        </tr>
      </thead>
      <tbody>
        <tr>
          <td
            class="sc-bjfHbI hnoYfZ"
          >
            Piso 1
          </td>
          <td
            class="sc-bjfHbI hnoYfZ"
          >
            <span
              style="color: green;"
            >
              100000
               €
            </span>
          </td>
          <td
            class="sc-bjfHbI hnoYfZ"
          >
            garaje
          </td>
        </tr>
        <tr>
          <td
            class="sc-bjfHbI hnoYfZ"
          >
            Piso 2
          </td>
          <td
            class="sc-bjfHbI hnoYfZ"
          >
            <span
              style="color: red;"
            >
              838000
               €
            </span>
          </td>
          <td
            class="sc-bjfHbI hnoYfZ"
          >
            chalets
          </td>
        </tr>
      </tbody>
    </table>
    <div
      class="sc-dIfARi cJZQWb"
    >
      <div
        class="sc-bcXHqe sc-idXgbr leBCrZ pCbZD"
      >
        <svg
          class="sc-pyfCe"
          height="20"
          width="20"
        >
          <polygon
            fill="#333"
            points="20,0 20,20 0,10"
          />
          Sorry, your browser does not support inline SVG.
        </svg>
        Pagina 
        1
         de 
        1
         
        <svg
          class="sc-pyfCe"
          height="20"
          width="20"
        >
          <polygon
            fill="#333"
            points="0,0 0,20 20,10"
          />
          Sorry, your browser does not support inline SVG.
        </svg>
      </div>
      <select
        id="pet-select"
        name="pets"
      >
        <option
          value="10"
        >
          10
        </option>
        <option
          value="25"
        >
          25
        </option>
        <option
          value="50"
        >
          50
        </option>
      </select>
    </div>
  </div>
</div>
`;
